cmd/apiserver-watcher/gcp_handler.go

```diff
@@ -0,0 +1,197 @@
+	}
+	// try to install the iptables rules for the VIPs
+	if err := gcp.syncRulesOnce(); err != nil {
+		return fmt.Errorf("error syncing iptables rules: %v", err)
+	}
+
+	ip := net.ParseIP(gcp.vip)
+	// bz1930457 delete stale conntrack entries to the apiserver
+	filter := &netlink.ConntrackFilter{}
+	filter.AddIP(netlink.ConntrackNatSrcIP, ip)
+	if _, err := netlink.ConntrackDeleteFilter(netlink.ConntrackTable, netlink.FAMILY_V4, filter); err != nil {
+		glog.V(4).Infof("Error deleting conntrack entries for %s", gcp.vip)
+	}
+
+	go gcp.syncRulesUntil()
+	return nil
+
+}
+
+// https://github.com/openshift/installer/blob/master/data/data/gcp/network/lb-private.tf
+// healthy_threshold   = 3
+// unhealthy_threshold = 3
+// check_interval_sec  = 2
+// timeout_sec         = 2
+
+func (gcp *gcpHandler) successThreshold() int {
+	return 1
+}
+
+func (gcp *gcpHandler) failureThreshold() int {
+	// LB = 6 seconds, plus 10 seconds for propagation
+	return 8
+}
+
+// syncRulesOnce syncs ip masquerade rules
+func (gcp *gcpHandler) syncRulesOnce() error {
+	gcpChains := []string{gcpMasqChainName, gcpLocalMasqChainName}
+	for _, gcpChan := range gcpChains {
+		exists, err := gcp.ipt.ChainExists("nat", gcpChan)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			if err := gcp.ipt.NewChain("nat", gcpChan); err != nil {
+				return err
+			}
+		}
+	}
+
+	// traffic coming from outside is redirected to the local apiserver
+	if err := gcp.ipt.AppendUnique("nat", "PREROUTING", "-m", "comment", "--comment", "gcp LB vip DNAT", "-j", gcpMasqChainName); err != nil {
+		return err
+	}
+	// traffic coming from inside is redirected to the local apiserver
+	if err := gcp.ipt.AppendUnique("nat", "OUTPUT", "-m", "comment", "--comment", "gcp LB vip DNAT for local clients", "-j", gcpLocalMasqChainName); err != nil {
+		return err
+	}
+	// Need this so that existing flows (with an entry in conntrack) continue,
+	// even if the iptables rule is removed
+	if err := gcp.ipt.AppendUnique("filter", "INPUT", "-m", "comment", "--comment", "gcp LB vip existing",
+		"-m", "addrtype", "!", "--dst-type", "LOCAL", "-m", "state", "--state", "ESTABLISHED,RELATED", "-j", "ACCEPT"); err != nil {
+		return err
+	}
+	if err := gcp.ipt.AppendUnique("filter", "OUTPUT", "-m", "comment", "--comment", "gcp LB vip existing",
+		"-m", "addrtype", "!", "--dst-type", "LOCAL", "-m", "state", "--state", "ESTABLISHED,RELATED", "-j", "ACCEPT"); err != nil {
+		return err
+	}
+
+	// Add LoadBalancer VIPs rule
+	glog.Infof("Adding iptables VIP rules for: %v", gcp.vip)
+	if err := gcp.ipt.AppendUnique("nat", gcpMasqChainName, "--dst", gcp.vip, "-j", "REDIRECT"); err != nil {
+		return err
+	}
+	if err := gcp.ipt.AppendUnique("nat", gcpLocalMasqChainName, "--dst", gcp.vip, "-j", "REDIRECT"); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// syncRulesUntil syncs the iptables rules until it receives the stop signal
+func (gcp *gcpHandler) syncRulesUntil() {
+	ticker := time.NewTicker(30 * time.Second)
```

Comment:
why do you think 30s is correct? Don't we sync much faster in the bash script (we have to!) ?

Reply:
the bash script has to sync faster because is not able to react to apiserver changes, it depends on this pod, to writes the api status file, and the insert the rules.
This scripts reacts intimidately to the apiserver status so the rules are programmed immediately., this cycle is just to be sure that nobody delete our rules